text/2022-05-12-multi-cluster-support.md

```diff
@@ -0,0 +1,335 @@
+# Multi Cluster Support
+
+## Summary
+
+As the multi-cluster injection is a huge feature, and related with the full
+lifecycle of Chaos Mesh, this RFC will be organized in a different way. It will
+describe the design through the Deploy -> Usage -> Authentication ->
+Architecture, and tell you the reason behind the design and other considerations
+in every separated section in a Q&A form.
+
+## Deploy
+
+A new cluster scope resource RemoteCluster will be used to install and manage
+the chaos-mesh installation on another cluster. This resource will be
+responsible to install, configure, upgrade and uninstall the cluster.
+
+```yaml
+apiVersion: chaos-mesh.org/v1alpha1
+kind: RemoteCluster
+metadata:
+ name: cluster-xxxxxxx
+spec:
+ namespace: "chaos-mesh"
+ kubeConfig:
+   secretRef:
+     name: “cluster-xxx-kubeconfig”
+   key: xxx
+ configOverride:
+   chaosDaemon:
+     runtime: containerd
+status:
+ currentVersion: "v2.1.0"
+ conditions:
+ - type: Installed
+   status: True
+ - type: Ready
+   status: True
+```
+
+The `Spec` defines the cluster which we want to install the chaos mesh, and the
+version of chaos mesh. It will install the same version of Chaos Mesh as the
+controller’s, which means if the version of controller is v2.2.0, it will
+install or upgrade the chaos mesh to v2.2.0 in the target cluster. It will only
+support the same minor version (e.g. parent cluster "v2.2.0" could control
+remote cluster "v2.2.1"). The `configOverride` allows to override the global
```

Comment:
> it will
install or upgrade the chaos mesh to v2.2.0 in the target cluster

Do we have any options that can control to keep the same version of parent cluster or not? The reason for controlling the installation of Chaos Mesh is the consideration of a SaaS Provider, but consider if there are massive clusters, the upgrade operation among all clusters looks heavy and dangerous (I mean the compatibility between versions). So I think upgrading all clusters automatically is not a good way. 🤔  

>  It will only
support the same minor version (e.g. parent cluster "v2.2.0" could control
remote cluster "v2.2.1").

What will happen if the versions are not matched?

Reply:
> The upgrade operation among all clusters looks heavy and dangerous (I mean the compatibility between versions)

I've though about adding a `version` field to the spec and status to manage the update, if the version in spec is greater than the one in the status, it will be upgraded automatically. However, I found it a little redundant, because the user always has to upgrade the version, or he will not be able to inject chaos into that cluster (as we'll refuse to inject into a cluster whose version doesn't match the parent one's).

As one resource only controls one cluster, upgrading all cluster seems to be heavy and dangerous (and the error must occur if the number is big), but upgrading one seems fine, (if the chaos mesh on remote cluster itself can handle the compatibility and availability during upgrading).

Reply:
@YangKeao Maybe I didn't clarify my question, what will happen when upgrading the parent cluster? Will the remote clusters be upgraded at the same time? If not, how to keep the same version with parent cluster?

And I have another new question 🤣, what will happen if the version of parent cluster is lower than the version of remote cluster? (case 1: the same minor version, case 2: the different minor version)

Reply:
@YangKeao Maybe I didn't clarify my question, what will happen when upgrading the parent cluster? Will the remote clusters be upgraded at the same time? If not, how to keep the same version with parent cluster?

And I have another new question 🤣, what will happen if the version of parent cluster is lower than the version of remote cluster? (case 1: the same minor version, case 2: the different minor version)

Reply:
> Will the remote clusters be upgraded at the same time?

Yes, because even it's not automatically done, the user will have to do it manually sometime.

> What will happen if the version of parent cluster is lower than the version of remote cluster? (case 1: the same minor version, case 2: the different minor version)

For same minor version, **I hope** it works fine. For different minor versions, it will refuse to work, as the target cluster version will be recorded in the status of `RemoteCluster`.

Reply:
Maybe we could add some annotations or something to allow users to stop the automatically upgrade ? :thinking: .

Reply:
> provide an option to choose upgrading automatically or not (just like the above comment said), and warn users this operation has some risk

So the only problem left about this topic is the default value? Should we upgrade automatically by default? What's your opinion?

> throw error to users when they want to inject experiments to the 'illegal' remote cluster, let them upgrade the remote cluster

Yes. Throwing error is expected, and we can even do this check in the webhook to tell the user the remote cluster version mismatched, and the user can solve through upgrading the remote cluster.

> If I understand correctly, it means the remote cluster will not be downgraded, right?

Yes.

Reply:
> So the only problem left about this topic is the default value? Should we upgrade automatically by default? What's your opinion?

Just my opinion, as 'upgrade automatically' is not a 'safe' operation at this moment, I prefer to let users upgrade manually as default, if they think upgrade manually is a heavy thing and they could accept the risk, then they could 'upgrade automatically'. Maybe we could ask more people for different thoughts.